libs/libc/unistd/lib_getoptvars.c

```diff
@@ -0,0 +1,166 @@
+
+#include <nuttx/config.h>
+
+#include <unistd.h>
+#include <pthread.h>
+
+#include "unistd.h"
+#include "libc.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define INVALID_KEY    CONFIG_TLS_NELEM
+#define VALID_KEY(key) ((key) >= 0 && key < CONFIG_TLS_NELEM)
+
+/* Configuration.  Currently no thread-specific data destructors. */
+
+#undef  HAVE_DESTRUCTOR
+#define HAVE_CLEANUP   1
+#define HAVE_TLSVARS   1
+
+/* Check if we have some way to avoid a memory leak */
+
+#if defined(CONFIG_DISABLE_PTHREAD) || !defined(CONFIG_PTHREAD_CLEANUP)
+#  undef HAVE_CLEANUP
+#endif
+
+/* Don't use TLS if we cannot avoid the memory leak or if TLS is disabled */
+
+#if (!defined(HAVE_DESTRUCTOR) && !defined(HAVE_CLEANUP)) || \
+    CONFIG_TLS_NELEM == 0
+#  undef HAVE_TLSVARS
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+static pthread_key_t g_getopt_key     = (pthread_key_t)INVALID_KEY;
+static pthread_mutex_t g_getopt_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+#else
+/* Fallback to avoid memory leak when there is no destructor and no cleanup
+ * or when there is no TLS.
+ */
+
+FAR struct getopt_s g_getopt_vars =
+{
+  NULL,
+  0,
+  1,
+  '?'
+};
+#endif
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+#ifdef HAVE_TLSVARS
+void _getopt_destructor(FAR void *getoptvars)
+{
+  lib_free(getoptvars);
+}
+#endif
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getoptvars
+ *
+ * Description:
+ *   Returns a pointer to to the thread-specific getopt() data.
+ *
+ ****************************************************************************/
+
+FAR struct getopt_s *getoptvars(void)
+{
+#ifdef HAVE_TLSVARS
+  FAR struct getopt_s *go;
+  int ret;
+
+  ret = pthread_mutex_lock(&g_getopt_mutex);
+  DEBUGASSERT(ret == OK);
+  UNUSED(ret);
+
+  /* Check if a key has been assigned */
+
+  if (!VALID_KEY(g_getopt_key))
+    {
+      ret = pthread_key_create(&g_getopt_key, _getopt_destructor);
+      DEBUGASSERT(ret == OK);
+      UNUSED(ret);
+    }
+
+  /* Get the structure of getopt() variables using the key. */
+
+  go = (FAR struct getopt_s *)pthread_getspecific(g_getopt_key);
+  if (go == NULL)
+    {
+      /* The instance has not yet been created.  Allocate getopt()
+       * it now.
+       */
+
+      go = (FAR struct getopt_s *)lib_malloc(sizeof(struct getopt_s));
```

Comment:
memory leak here, no body free this memory after the task exit, we need first finish:
https://github.com/apache/incubator-nuttx/issues/1263
https://github.com/apache/incubator-nuttx/tree/feature/pthread-user

Reply:
This should not be a memory leak.  This function is called when the thread exits.  PR #1263 has to do with a security error in PROTECTED and KERNEL modes.  In those modes, the callbacks are made in supervisor mode which is a problem.

However, the callbacks to work.  The following is called (I have verified that) and there is no memory leak.

```
 91 #ifdef HAVE_TLSVARS
 92 void _getopt_destructor(FAR void *getoptvars)
 93 {
 94   lib_free(getoptvars);
 95 }
 96 #endif
```

Wait, I will re-verify.

Reply:
But I wonder if we could not release the memory earlier than waiting for the thread to exit.  It is difficult to know exactly when the multi-step getopt() sequence is finished.  But certainly is finished where ever binitialized is set false and could be freed at all of the points (although we would also have to deal with undoing the thread exit hooks which I am not sure is possible).

Reply:
pthread_keycreate never save the destruct callback:
https://github.com/apache/incubator-nuttx/blob/master/libs/libc/pthread/pthread_keycreate.c#L75-L97
How the pthread can find the callback and then call it?

Reply:
Ok, the destroy is trigger by pthread_cleanup_push, but I am not sure whether it is the right approach to trigger the callback. The implementation require we:

1. Enable thread local storage
2. Enable thread clean up stack

Both options are disabled by default? The change will break most project.

Reply:
> How can the exit path find the callback and then call it?

It can't.  pthread data destructors have never been implemented.  @antmerlino was planning on doing this on https://github.com/apache/incubator-nuttx/tree/feature/pthread-user, but that effort lost focus.

This is handled in libs/libc/unistd/lib_getoptvars.c.  Basically, it uses pthread_cleanup() instead of the data destructor.  That works just fine in this case:

```
 42 #undef  HAVE_DESTRUCTOR

 91 #ifdef HAVE_TLSVARS
 92 void _getopt_destructor(FAR void *getoptvars)
 93 {
 94   lib_free(getoptvars);
 95 }
 96 #endif

124       ret = pthread_key_create(&g_getopt_key, _getopt_destructor);

155 #if !defined(HAVE_DESTRUCTOR) && defined(CONFIG_PTHREAD_CLEANUP)
156       if (ret == OK)
157         {
158           /* Free the thread-specific data when the thread exits */
159
160           pthread_cleanup_push(_getopt_destructor, go);
161         }
162 #endif

```
The second parameter to pthread_key_create() at line 96 does nothing as you note.  That actually deallocation is setup by pthread_cleanup_push() at line 160 when the pthread exits.  That is sufficient in this case

Reply:
> Both options are disabled by default? The change will break almost project.



> Both options are disabled by default? The change will break almost project.

No, that case is handled.  It is not a serious problem.  It will simply revert to the previous behavior:

```
 52 /* Don't use TLS if we cannot avoid the memory leak or if TLS is disabled.
 53  * Furthermore, TLS is not needed in the KERNEL build because it naturally
 54  * supports per-process data.
 55  */
 56
 57 #if !defined(HAVE_DESTRUCTOR) && !defined(HAVE_CLEANUP)
 58 #  undef HAVE_TLSVARS
 59 #endif
 60
 61 #ifdef CONFIG_BUILD_KERNEL
 62 #  undef HAVE_TLSVARS
 63 #endif

 69 #ifdef HAVE_TLSVARS
 73 #else
 74 /* Fallback to avoid memory leak when there is no destructor and no cleanup
 75  * or when there is no TLS.
 76  */
 77
 78 FAR struct getopt_s g_getopt_vars =
 79 {
 80   NULL,
 81   0,
 82   1,
 83   '?',
 84   NULL,
 85   false
 86 };
 87 #endif

```

Reply:
NOTE also that TLS is now unconditionally enabled.  CONFIG_TLS_NELEM is not permitted to take the value zero any long and all conditional logic based on CONFIG_TLS_NELEM has been removed.

There has to be some callback to free the memory or else it will again fallback to the legacy behavior.

I think these configuration dependencies could be improved, but I am not sure how to do that now.

Reply:
Implementing the destructors the correct, secure way is complex.  Basically, pthread_exit() has to exit in user space, not in the kernel.  Then that user space logic can perform the pthread_cleanup_pop() and the call the data destructors.  Then it can call into the kernel to finish the pthread_exit()

That means also moving the pthread startup function to user space (since it calls pthread_exit()).

Those user-space addresses are not known in by the kernel in PROTECTED and KERNEL build modes and have to be passed to the OS through extensions to the kernel interface.  That is the source of the complexity of doing things the correct way instead of the easy way.  I am not sure if I am feeling ambitious enough to do that now.

Reply:
> Implementing the destructors the correct, secure way is complex. Basically, pthread_exit() has to to called in user space, not from kernel space. Then that user space logic can perform the pthread_cleanup_pop() and the call the data destructors securely. Then it can call into the kernel to finish the pthread_exit() in supervisor mode.
> 

Yes, but the major work is moving code from sched/pthread to libc/pthread.

> That means also moving the pthread startup function to user space (since it calls pthread_exit()).
> 
> Those user-space addresses are not known in by the kernel in PROTECTED and KERNEL build modes and have to be passed to the OS through extensions to the kernel interface. That is the source of the complexity of doing things the correct way instead of the easy way. I am not sure if I am feeling ambitious enough to do that now.
> 

Yes, we don't need introuce the new hook here, since userspace_s already has these callbacks:
```
struct userspace_s
{
  /* Task/thread startup routines */

  CODE void (*task_startup)(main_t entrypt, int argc, FAR char *argv[]);
#ifndef CONFIG_DISABLE_PTHREAD
  CODE void (*pthread_startup)(pthread_startroutine_t entrypt,
    pthread_addr_t arg);
#endif
};
```

> We have not been good at doing large extended projects on branches. Perhaps we could take this a step at a time, merging to master after each step:
> 
> 1. Move pthread_start to user space
> 2. Move pthread_exit to user space
> 3. Implement pthread specific data destructors.
> 
> That would not be as intimidating.

Yes, it is huge reconstructure. Or we can at least implement the pthread destructor same as how at_exit work.